bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java

```diff
@@ -0,0 +1,530 @@
+    }
+
+    public boolean hasStatusValChanged(String valueName) throws GreeException {
+        if (!prevStatusResponsePackGson.isPresent()) {
+            return true; // update value if there is no previous one
+        }
+        // Find the valueName in the Current Status object
+        List<String> currcolList = Arrays.asList(statusResponseGson.get().packJson.cols);
+        List<Integer> currvalList = Arrays.asList(statusResponseGson.get().packJson.dat);
+        int currvalueArrayposition = currcolList.indexOf(valueName);
+        if (currvalueArrayposition == -1) {
+            throw new GreeException("Unable to decode device status");
+        }
+
+        // Find the valueName in the Previous Status object
+        List<String> prevcolList = Arrays.asList(prevStatusResponsePackGson.get().cols);
+        List<Integer> prevvalList = Arrays.asList(prevStatusResponsePackGson.get().dat);
+        int prevvalueArrayposition = prevcolList.indexOf(valueName);
+        if (prevvalueArrayposition == -1) {
+            throw new GreeException("Unable to get status value");
+        }
+
+        // Finally Compare the values
+        return currvalList.get(currvalueArrayposition) != prevvalList.get(prevvalueArrayposition);
+    }
+
+    protected void executeCommand(DatagramSocket clientSocket, Map<String, Integer> parameters) throws GreeException {
+        // Only allow this to happen if this device has been bound
+        if (!getIsBound()) {
+            throw new GreeException("Device is not bound!");
+        }
+
+        try {
+            // Convert the parameter map values to arrays
+            String[] keyArray = parameters.keySet().toArray(new String[0]);
+            Integer[] valueArray = parameters.values().toArray(new Integer[0]);
+
+            // Prep the Command Request pack
+            GreeExecuteCommandPackDTO execCmdPackGson = new GreeExecuteCommandPackDTO();
+            execCmdPackGson.opt = keyArray;
+            execCmdPackGson.p = valueArray;
+            execCmdPackGson.t = GREE_CMDT_CMD;
+            String execCmdPackStr = gson.toJson(execCmdPackGson);
+
+            // Now encrypt and send the Command Request pack
+            String encryptedCommandReqPacket = GreeCryptoUtil.encryptPack(getKey().getBytes(), execCmdPackStr);
+            DatagramPacket sendPacket = createPackRequest(0,
+                    new String(encryptedCommandReqPacket.getBytes(), StandardCharsets.UTF_8));
+            clientSocket.send(sendPacket);
+
+            GreeExecResponseDTO execResponseGson = gson.fromJson(receiveResponse(clientSocket),
+                    GreeExecResponseDTO.class);
+            execResponseGson.decryptedPack = GreeCryptoUtil.decryptPack(this.getKey().getBytes(),
+                    execResponseGson.pack);
+
+            // Create the JSON to hold the response values
+            execResponseGson.packJson = gson.fromJson(new JsonReader(new StringReader(execResponseGson.decryptedPack)),
```

Comment:
There are a lot more places that you are using the JsonReader instead of the string directly, so I suspect you are doing it intentionally. If you aren't then change them all to use the string directly as per my earlier comment.

Reply:
this is from the old code; I did already changed it, but then users reported exceptions. I don't understand why, but after reverting the change, exception was gone

Reply:
Do you know what the exception was?

Reply:
I was able to debug it, the packet is padded with 0x00 bytes, refactored the response handling to trim the string, then it works